### ISSUE TITLE
Issues on VGG16_fc_reduced caffemodel

### ISSUE THREAD

**Issue by stoneyang** [weiliu89/caffe#365]
Hi, Wei @weiliu89 :

Thanks for your awesome code! 

I am working on porting the work of SSD to equip base net other than the specialized VGG16 in your original paper. One of the most important modification on VGG16 is on `pool5` and `fc6`: 

- `pool5`:
  - `kernel_size` is changed from `2` to `3`

  - `stride` from `2` to `1`, and

  - `pad` is assigned to `1`;

- `fc6`: 
  - `kernel_size` is assigned to `3`

  - `dilation` is `6`

  - `pad` is `6`, and

  - `num_output` is 1024

According to the expression defined dilation in [`conv_layer.cpp`](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/conv_layer.cpp#L17), if one need to keep the size of the feature map from reducing (am I right?), i.e., `14` after modified `pool5` in the VGG16 on ImageNet case, the following condition should be met: 

`floor( ( 14 + 2 * pad - ( dilation * ( filter_size - 1 ) + 1 )  ) / 1 + 1 ) = 14`

that is, `pad = dilation`, which is undetermined. So, can we set them both to other values?

Looking forward to your reply. 

Regards.

**Comment by weiliu89** [comment]
I think you have to first figure out the `dilation` value after all the changes. For example, as you pointed out, the effective `kernel_size` of fc6 is 14. And I only subsample 3x3 values from the original 7x7 kernel. Thus, the dilation should be 6.

Then you can compute pad (i.e. [here](https://github.com/weiliu89/caffe/blob/ssd/python/caffe/model_libs.py#L286)) to make sure after convolution the feature map has the same size (if that is what you want).

**Comment by qingxu20170103** [comment]
Wei, Thanks for explain! It also answered my question.

I also want to ask why the fc6 has 1024 outputs instead of the original 4096 outputs in vgg-16. Did you also sample the 1024 out of the original 4096 in vgg16 model? Thanks.

**Comment by weiliu89** [comment]
@qingxu20170103 Yes. I subsample 1024 from 4096 (i.e. one per 4 feature 0:4:4096)

**Comment by qingxu20170103** [comment]
@weiliu89 Thank you very much!

**Comment by sergii-bond** [comment]
Just an additional comment to explain formula from _model_libs.py_

`pad = int((kernel_size + (dilation - 1) * (kernel_size - 1)) - 1) / 2`

In _conv_layer.cpp_ the output dimension is calculated as:
```
    const int kernel_extent = dilation_data[i] * (kernel_shape_data[i] - 1) + 1;
    const int output_dim = (input_dim + 2 * pad_data[i] - kernel_extent)
        / stride_data[i] + 1;
```

After little algebra it can be seen that 'pad' expression makes output_dim equal to input_dim.
